### PR TITLE
Add BASNet Segmentation Model (#2006)

### LINKED ISSUE #2006: BASNet model
Add the BASNet segmentation model to KerasCV.

A keras.io [BASNet example](https://keras.io/examples/vision/basnet_segmentation/) already exists, based on this [BASNet implementation](https://github.com/hamidriasat/BASNet/tree/basnet_keras), both from [Hamid Ali](https://github.com/hamidriasat).

The API should follow that of the existing [DeepLabV3 model](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/segmentation/deeplab_v3_plus/deeplab_v3_plus.py).

### PR DESCRIPTION
**Pull request by hamidriasat**
# What does this PR do?
Adding [BASNet](https://arxiv.org/abs/2101.04704) model.

**Overview:**
This pull request addresses the assigned task (#2006) to integrate the BASNet segmentation model into KerasCV. The additions include the implementation of the BASNet model architecture and the inclusion of relevant test cases.

**Weights and Future Plan:**
I've included the BASNet model and its tests in this submission. Once they're reviewed and merged, I'll follow up with another PR to add the weights.

Fixes #2006

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case. [BASNet model](https://github.com/keras-team/keras-cv/issues/2006)
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

@divyashreepathihalli @tirthasheshpatel @martin-gorner @jbischof

### CODE REVIEW

File: keras_cv/models/segmentation/basnet/basnet_test.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import tensorflow as tf
+
+from keras_cv.backend import keras
+from keras_cv.backend import ops
+from keras_cv.backend.config import keras_3
+from keras_cv.models import BASNet
+from keras_cv.models import ResNet34Backbone
+from keras_cv.tests.test_case import TestCase
+
+
```
Comment (divyashreepathihalli) [inline keras_cv/models/segmentation/basnet/basnet_test.py:28]: lets add a test for forward pass and also a test for presets - example [here](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/segmentation/deeplab_v3_plus/deeplab_v3_plus_test.py)

**Review by divyashreepathihalli — COMMENTED** [review]
This is awesome!! Thank you for this contribution. I just have one more comment on increasing the test coverage.

**Comment by hamidriasat** [comment]
@divyashreepathihalli  Thanks! I'm delighted that you find it useful.

**Recent commit**:

1. Improved test coverage as requested.
2. Fixed linter issues.
3. Added Keras3 support, resolving the mentioned error.
4. [Here](https://colab.research.google.com/drive/1Op1MQ2OPMbmADXCdT2IKuWHbr-rmPm_h?usp=sharing) is a colab script which supports model training on paper used dataset. I reused the code from one of my own contributions from [here](https://keras.io/examples/vision/basnet_segmentation/). I skipped TPU training because for that the data needed to be in GCP bucket which I do not have access right now. You can change the hyperparameters based on available resources for training.

Now, before merging with the master branch, all tests have passed successfully. Currently, one test `test_with_model_preset_forward_pass` is giving an error due to a recent kaggle_handle change.

Could you please clarify if you are going to upload the model to Kaggle and share the Kaggle handle link, or how this will work?

**Review by divyashreepathihalli — APPROVED** [review]

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
I will need to upload the weights to Kaggle. Lets disable to preset tests for now.

**Comment by hamidriasat** [comment]
@divyashreepathihalli Preset test disabled for now.

**Review by divyashreepathihalli — APPROVED** [review]

**Comment by divyashreepathihalli** [comment]
Thank you for the awesome contribution. The PR is merged. We will upload the weights to Kaggle and add the presets

**Comment by divyashreepathihalli** [comment]
@hamidriasat  the training is resulting in this error
```
WARNING:tensorflow:Gradients do not exist for variables ['conv1_conv/kernel:0', 'conv1_bn/gamma:0', 'conv1_bn/beta:0'] when minimizing the loss.
```

**Pull request merged**